### ISSUE TITLE
Support immutable entities

### ISSUE THREAD

**Issue by doanduyhai** [doanduyhai/Achilles#264]
- add an `immutable` attribute on the `@Table` annotation
- introspection to use non default public constructor (probably another impl of BeanFactory)
- if multiple non default public constructors are found, pick one with `@DefaultConstructor`. If not found, raise exception
- forbid setters on fields passed to the non default public constructor
- on fields that are **not** passed to non default public constructor, ensure setters are there
- check that all getters are present

Did I miss something @bric3 @KeepItSimpleStupid

**Comment by KeepItSimpleStupid** [comment]
@doanduyhai, I would say that the 2 following points are not the responsability of Achilles :
- forbid setters on fields passed to the non default public constructor
- on fields that are not passed to non default public constructor, ensure setters are there

I also didn't get the need of this new attribute "immutable" on @Table? 

IMO, the idea would be to put everything in place to promote immutable entities (by using constructor on deserialization and getters on serializations), to be clear in the doc that's the model Achilles favor, but not to force the user in any manner 😉

**Comment by KeepItSimpleStupid** [comment]
I would also rename the annotation @DefaultConstructor to @EntityCreator (similar to Jackson's @JSONCreator ☺️

**Comment by bric3** [comment]
> * introspection to use non default public constructor (probably another impl of BeanFactory)

Note that if the code is generated, the persistence manager maybe able to instantiate without introspection, the annotation processor could just generate the code that will use the annotated constructor.

@KeepItSimpleStupid `immutable` would indicate how code generation performs, e.g. 

* `immutable = false`, current behaviour and code gen
* `immutable = true`, code gen will look for a **single annotated constructor**, will generate getters only, and use default constructor.

And this allow users to opt-in change (before making default or not)

**Comment by doanduyhai** [comment]
1. Ok for `@EntityCreator`
2. With regard to constructors
    1. when a single non-default constructor is found, use it 
    2. when multiple non-default constructors are found, use the one with `@EntityCreator` or exception if annotation not found
3. For each **mapped** fields
    1. if the field is present in the `@EntityCreator` constructor
        1. if this field has `public final` modifiers, no modifier check (or maybe forbid getter/setter @bric3 @KeepItSimpleStupid  ?)
        2. if this field has `private` modifier, ensure getter is present
    2. if the field is **not** present in the `@EntityCreator` constructor, ensure getters/setters are present

**Comment by bric3** [comment]
> * if the field is not present in the @EntityCreator constructor, ensure getters/setters are present

Why enforcing the presence of getters/setters, this is wrong, especially the setters. Maybe the dev don't want to map some fields.

**Comment by KeepItSimpleStupid** [comment]
For 3. let's decompose in serialization/deserialization scenarios :

1. Serialization :
i. Find all the mapped fields (all the class fields and getter methods annotated @Column)
ii. Verify there's no field mapped twice
iii. Insert

2. Deserialization :
i. Determine the deserialization strategy to use : Constructor or Setters
if there's at least one non default constructor, use Constructor strategy. If not, use the Setter strategy (the current one used by Achilles 5.x in fact)

ii. If Constructor strategy, determine the constructor to use :
if only one non default constructor, use this one. If multiples, use the one with @JsonCreator. If multiples, but none with @JsonCreator, refuse to start / compile.

iii. One you have the constructor to use, you should map your Cassandra table fields to the Constructor arguments. You can make it mandatory to annotate (again) the constructor arguments with `@Column` like Brice suggested (that's the strategy used by Jackson), or you can rely on obtaining parameter names at runtime but you have to force your user to compile with `-parameter` to access it.

Note, this is not mandatory that all the fields mapped during Serialization are the same than the one during Deserialization.

**Comment by bric3** [comment]
@KeepItSimpleStupid I'd add also that the code should chose either constructor strategy or setter strategy, but not constructor first then setter, it's wrong for entities.

**Comment by doanduyhai** [comment]
So first remark, we don't need `@Column` on constructor arg, annotation processor can access argument name. But then we may have issue e.g.:

```java
public final String name;
public final int age;
...
@EntityCreator
public User(String fullName, int age) {
   this.name = fullname;
   this.age = age;
}
```

As you can see, without `@Column` on fullName, we cannot know that it is related to `public final name`, unless we explicitly require field name enforcement for the user.

Second issue: I strongly disagree with _this is not mandatory that all the fields mapped during Serialization are the same than the ones during Deserialization_. You're introducing asymmetric behavior between INSERT/SELECT e.g. the data you get at SELECT may not be the same as the data you INSERTED. It is confusing to average users and can lead to unexpected behavior.

Last remark, the rules are getting too complicated because there are too many cases. It's an indication of a bad design or incorrect abstraction level. Let's do something less powerful but much simpler to follow the principle of _less surprise_ for the users.

1. With regard to constructors
    1. when a single non-default constructor is found, use it
    2. when multiple non-default constructors are found, use the one with `@EntityCreator` or exception if annotation not found
2. For each **mapped field** (having `@Column` or not `@Transient` if implicit mapping strategy is used)
    1. Require them to be present in the `@EntityCreator` constructor with explicit `@Column` annotation to match the param with the correct column
    2. Require them to have `public final` modifiers and not getter/setter

In the end we'll have only 2 simple strategies:

* Java bean classic strategy (current behavior)
* pure immutable strategy (whenever a constructor with `@EntityCreator` is found)

This seems very restrictive but I prefer we do only one thing but correctly rather than trying to cover many distinct scenarios with potential for bugs.

**Comment by doanduyhai** [comment]
https://github.com/doanduyhai/Achilles/wiki/Achilles-Annotations#immutable

Released in **Achilles 5.3.0**
